### ISSUE TITLE
Rules

### ISSUE THREAD

**Issue by MoOx** [stylelint/stylelint#1]
_Vocabulary reference http://apps.workflower.fi/vocabs/css/en_

Updated list

---

# Rules

## General (tied to the stylesheet itself)

* [x] `no-missing-eof-newline`: Disallow missing end-of-file newline
* [x] `no-multiple-empty-lines`: Disallow multiple empty white lines
* [x] `no-eol-no-whitespace`: Disallow whitespace at the end of line
* [x] `indentation`: Specify whitespace for indentation
* [ ] `max-line-length`: Specify maximum line length

## String ([ref](http://www.w3.org/TR/css-cascade-3/#at-import))

* [x] `string-quotes`: `"double"|"single"` Specify quotes around strings
  * `"double"` - strings *must* be doubled quoted
  * `"single"` - strings *must* be single quoted

## Number ([ref](http://www.w3.org/TR/css3-values/#number))

* [x] `number-leading-zero`: `"always"|"never"` Require or disallow a leading zero before number
  * `"always"` - a number *must* have a leading zero
  * `"never"` - a number *must not* have a leading zero
* [x] `number-no-trailing-zeros`: Disallow trailing zeros (e.g. `5.0px`)
* [x] `number-zero-length-no-unit`: Disallow units for zero length values
* [x] `number-max-precision`: `int` Maximum number of digits after the "." in a number

## Function ([ref](http://www.w3.org/TR/css3-values/#functional-notation))

* [x] `function-comma-space-after`: `"always"|"never"`
* [x] `function-comma-space-before`: `"always"|"never"`
* [x] `function-parentheses-inside-space`: `"always"|"never"` Specify space inside parentheses (after opening, before closing)
* [x] `function-space-after`: `"always"|"never"` Specify space between the closing parenthesis of a function and the next value
* [x] `function-token-no-space`: Disallow space between the function's name and its opening parenthesis
* [x] `function-calc-no-unspaced-operator`: Disallow operators without space on both sides in `calc()`
* [x] `function-url-quotes`: `"double"`, `"single"`, `"none"` Specify quotes around URLs (URLs, unlike strings, can be unquoted)
  * `"double"` - URLs *must* be doubled quoted
  * `"single"` - URLs *must* be single quoted
  * `"none"` - URLs *must* be unquoted

## Color ([ref](http://dev.w3.org/csswg/css-color-4/))

* [x] `color-hex-length`: `"short"|"long"`
* [x] `color-hex-case`: `"lower"|"upper"`
* [x] `color-no-invalid-hex`: Disallow invalid hex
* [ ] `color-no-named`: Disallow named colors
* [ ] `color-no-hex`: Disallow hex colors
* [ ] `color-function-blacklist`: `array` Disallowed color functions e.g. `["rgba", "hsl", "hwb"]`
* [ ] `color-function-whitelist`: `array` Only allowed color functions e.g. `["gray", "color", "device-cmyk"]`

## :root

* [x] `root-no-standard-properties`: Disallow use of standard properties in :root

## Rule

* [x] `rule-no-single-line`: Disallow single line rule-sets
* [x] `rule-properties-order`: `1 dimension array`, `string` (preset) - `"alphabetical"` Order of properties within a rule
* [x] `rule-trailing-semicolon`: `"always"|"never"` Trailing semicolon at the end of a rule
  * `"always"` - there *must* be a trailing semicolon
  * `"never"` - there *must not* be a trailing semicolon
* [x] `rule-no-duplicate-properties`: Disallow duplicate properties within a rule
* [x] `rule-nested-empty-line-before`: Require or disallow an empty line before nested rules
* [ ] `rule-single-line-max-length`: `int` The maximum length of a single line rule
* [ ] `rule-single-line-max-declarations`: `int` The maximum number of declaration within a single line rule

## At-rule rules

* [x] `at-rule-no-vendor-prefix`: Disallow vendor prefixes in @rules
* [x] `at-rule-empty-line-before`: Require or disallow an empty line before at-rules 

## Custom media ([ref](http://dev.w3.org/csswg/mediaqueries-4/#custom-mq))

* [x] `custom-media-pattern`: `string` Preset or regex pattern to use to check names are matching a given pattern

## Media query

* [x] `media-query-parenthesis-inside-space`: `"always"|"never"`
* [x] `media-query-list-comma-space-after`: `"always"|"never"`
* [x] `media-query-list-comma-space-before`: `"always"|"never"`
* [x] `media-query-list-comma-newline-after`: `"always"|"never"`
* [x] `media-query-list-comma-newline-before`: `"always"|"never"`

## Media feature

* [x] `media-feature-colon-space-after`: `"always"|"never"`
* [x] `media-feature-colon-space-before`: `"always"|"never"`
* [x] `media-feature-range-operator-space-after`: `"always"|"never"`
* [x] `media-feature-range-operator-space-before`: `"always"|"never"`

## Media feature

* [x] `media-feature-colon-space-after`: `"always"|"never"`
* [x] `media-feature-colon-space-before`: `"always"|"never"`
* [x] `media-feature-range-operator-space-after`: `"always"|"never"`
* [x] `media-feature-range-operator-space-before`: `"always"|"never"`

## Media feature name

* [x] `media-feature-name-no-vendor-prefix`: `"always"|"never"` Disallow vendor prefixes for media feature names

## Selector

* [x] `selector-combinator-space-before`: `string` Specify what space should be used before and after combinator
* [x] `selector-combinator-space-after`: `string` Specify what space should be used before and after combinator
* [x] `selector-delimiter-space-after`: `"always"|"never"`
* [x] `selector-delimiter-space-before`: `"always"|"never"`
* [x] `selector-delimiter-newline-after`: `"always"|"never"`
* [x] `selector-delimiter-newline-before`: `"always"|"never"`
* [x] `selector-no-vendor-prefix`: Disallow vendor prefixes in selectors
* [x] `selector-pseudo-element-notation`:  Specify that pseudo element selectors need one or two
* [x] `selector-root-no-composition`: Disallow use of `:root` selector with others (in list, complex, or compound selectors)
* [x] `selector-no-id`: Disallow use of id in selectors e.g: `#id`
* [x] `selector-no-type`: Disallow use of type in selectors e.g: `div`
* [x] `selector-no-universal`: Disallow use of universal selector in selectors e.g: `*`
* [x] `selector-no-attribute`: Disallow use of attributes in selectors e.g: `[attr]`
* [x] `selector-no-combinator`: Disallow use of combinator in selectors
* [ ] `selector-no-qualified`: Disallow use of qualified selectors e.g: `div#id`
* [ ] `selector-no-pseudo-class`: Disallow use of pseudo-class in selectors
* [ ] `selector-no-pseudo-element`: Disallow use of pseudo-element in selectors
* [ ] `selector-no-delimiter`: Disallow use of delimiter in selectors
* [ ] `selector-pattern`: `string` regex (eg: `/\.([a-z]+-)?[A-Z][a-z]+(-[a-z][a-zA-Z]+)(--[a-z][a-zA-Z]+)/` to match .org-Block-eleMent--modifier

## Custom selector ([ref](http://dev.w3.org/csswg/css-extensions/#custom-selectors))

* [ ] `custom-selector-pattern`: `string` Regex pattern to use to check names are matching a given pattern

## Block

* [x] `block-no-empty`: Disallow empty blocks
* [x] `block-opening-brace-space-after`: `"always"|"never"|"always-single-line"|"never-single-line"|"always-multi-line"|"never-multi-line"`
* [x] `block-opening-brace-space-before`: `"always"|"never"|"always-single-line"|"never-single-line"|"always-multi-line"|"never-multi-line"`
* [x] `block-opening-brace-newline-after`: `"always"|"never"|"always-multi-line"|"never-multi-line"`
* [x] `block-opening-brace-newline-before`: `"always"|"never"|"always-single-line"|"never-single-line"|"always-multi-line"|"never-multi-line"`
* [x] `block-closing-brace-space-after`: `"always"|"never"|"always-single-line"`
* [x] `block-closing-brace-space-before`: `"always"|"never"|"always-single-line"|"never-single-line"|"always-multi-line"|"never-multi-line"`
* [x] `block-closing-brace-newline-after`: `"always"|"never"|"always-single-line"|"never-single-line"|"always-multi-line"|"never-multi-line"`
* [x] `block-closing-brace-newline-before`: `"always"|"never"|"always-multi-line"|"never-multi-line"`

## Nesting Block

* [x] `nesting-block-opening-brace-space-before`: `"always"|"never"|"always-single-line"|"never-single-line"|"always-multi-line"|"never-multi-line"`
* [x] `nesting-block-opening-brace-newline-before`: `"always"|"never"|"always-single-line"|"never-single-line"|"always-multi-line"|"never-multi-line"`

## Declaration rules

* [x] `declaration-no-important`: Disallow the use of !important
* [x] `declaration-bang-space-after`: `"always"|"never"`
* [x] `declaration-bang-space-before`: `"always"|"never"`
* [x] `declaration-colon-space-after`: `"always"|"never"`
* [x] `declaration-colon-space-before`: `"always"|"never"`
* [x] `declaration-semicolon-space-after`: ` "always"|"never"|"always-single-line"|"never-single-line"`
* [x] `declaration-semicolon-space-before`: ` "always"|"never"|"always-single-line"|"never-single-line"`
* [x] `declaration-semicolon-newline-after`: `"always"|"never"|"always-multi-line"|"never-multi-line"`
* [x] `declaration-semicolon-newline-before`: `"always"|"never"|"always-multi-line"|"never-multi-line"`

## Unit

* [ ] `unit-blacklist`: `array` Disallowed units e.g. `["px", "pt", "cm"]`
* [ ] `unit-whitelist`: `array` Only allowed units e.g. `["%", "rem", "em"]`

## Property

* [x] `property-no-vendor-prefix`: Disallow vendor prefixes in properties
* [x] `property-blacklist`: `array` Disallowed properties e.g. `["tranform", "background-size"]`
* [x] `property-whitelist`: `array` Only allowed properties e.g. `["height", "width", "font-size"]`
* [ ] `property-unit-blacklist`: `object` Disallowed units for specific properties e.g. `{"width": ["%"], "height": ["%"]}`
* [ ] `property-unit-whitelist`: `object` Only allowed units for specific properties e.g. `{"font-size": ["em", "rem"]}`

## Custom property

* [x] `custom-property-no-outside-root`: Disallow custom properties outside of `:root` 
* [x] `custom-property-pattern`: `string` Regex pattern to use to check names are matching a given pattern

## Value

* [x] `value-no-vendor-prefix`: Disallow vendor prefixes in values

## Value list

* [x] `value-list-comma-space-after`: ` "always"|"never"|"always-single-line"|"never-single-line"`
* [x] `value-list-comma-space-before`: ` "always"|"never"|"always-single-line"|"never-single-line"`
* [x] `value-list-comma-newline-after`: `"always"|"never"|"always-multi-line"|"never-multi-line"`
* [x] `value-list-comma-newline-before`: `"always"|"never"|"always-multi-line"|"never-multi-line"`

## Comment

* [x] `comment-space-inside`: `string` - `"always"|"never"
* [x] `comment-empty-line-before`: Require or disallow an empty line before comments 

---

# To think about Mistakes

- Float or absolutely positioned elements don't need display: block
- Absolutely positioned elements will ignore float

**Comment by jeddy3** [comment]
Excellent. It's agreed that the linter should be both:

* Pluggable.
* Non-opinionated.

And to do that:

1. *Every* rule should be standalone. 
2. *Every* rule should be turned off by default.

> What about "simply" create rules with complex options ?

Standalone rules with complex options could work! For example, I came across these two CSS rules from the [SCSS Styleguide](https://github.com/magnetis/styleguide/blob/master/Stylesheet.md) that @rafaelrinaldi maintains:

* Use `pt` units to declare `letter-spacing` values ([ref](https://github.com/magnetis/styleguide/blob/master/Stylesheet.md#letter-spacing)).
* Do not add units for `line-height` values ([ref](https://github.com/magnetis/styleguide/blob/master/Stylesheet.md#line-height)).

I originally thought that these would be perfect candidates for @rafaelrinaldi to write and plug in two of his own rules, which would be very narrowly-focused and have simple options. Something like:

* `value-letter-spacing-unit-points-only: bool`
* `value-line-height-unitless: bool`

But I then realised that a `*-blacklist` and `*-whitelist` rule, similar to `value-unit-blacklist` and  `value-unit-whitelist` but with the unit and property in the object swapped around to make it work, would be able to handle *all* these types of rules by using their complex options e.g.

* `property-unit-blacklist:  Object ({"width": ["%"], "height": ["%"]})`
* `property-unit-whitelist:  Object ({"font-size": ["em", "rem"]})`

So, for @rafaelrinaldi 's  styleguide it would be:

```js
property-unit-whitelist: {
  "letter-spacing": ["pt"], // i.e. letter-spacing property must only have "pt" units
  "line-height": [] // i.e. line-height property must not have any units (as nothing is whitelisted)
}
```

Up until then, I hadn't realised the power of your white/blacklist rules!

I guess it's all about finding a balance between very narrowly-focused rules and broader, but more powerful, ones. I think choosing which approach applies to which set of rules will need to happen on a case-by-case basis. For example, perhaps complex options *are* needed for things like `indentation` and `*-black/whitelist`, while simple options are perfect for things like `eof-newline`,`rule-set-no-empty`, and `selector-attribute-quotes`.

What are everyone else's thoughts on this? Do complex options, where needed, feel like the right way to go? 

The [esformatter](https://github.com/millermedeiros/esformatter/) that @rafaelrinaldi mentioned seems to take a [similar approach](https://github.com/millermedeiros/esformatter/blob/master/lib/preset/jquery.json) to complex options for `indentation` and `separation`.

---

@rafaelrinaldi  Can you write a little more about your thoughts on presets and plugins? 

Is it related to this [esformatter issue](https://github.com/millermedeiros/esformatter/issues/189) or how ESLint [approaches plugins](http://eslint.org/docs/developer-guide/working-with-plugins.html)? I just stumbled across them now, and it makes the preset route feel a bit shaky. Any input or advice?

---

@necolas Thanks, that's good to know! I think it's been a good exercise identifying lots and lots of rules, and seeing how far you can take thinking up a flexible linter before it becomes too much work :)

Now that plenty of rules have been jotted down, we can:

* Work out how the rules can be grouped and combined. This will also help us come up with a pretty consistent naming scheme. I think good progress has been made with this so far!
* Work out which rules might be easy to implement, and which might be hard.
* Think of low-hanging-fruit alternatives to the harder rules e.g instead of very elaborate rules for handling `separation` we could start with something like `no-multiple-empty-lines`, as we'd probably get good mileage out of that and it might be easier to implement.
* Do something like a poll where everyone gets to list, and prioritise, the 10 rules they'd find most useful. We can then focus on the rules that'll be useful to most number of people. From my perspective, the `declaration-block-properties-order` rule alone would help me and guys I work with write more consistent code for each other!

I think all that will feed into which, if any, of the idiosyncratic aspects could be removed from the idiomatic-css or suitcss styles. I'm hoping none as the guides are already very succinct! :)

---

Lastly, one more rule for the list?:

* `value-max-precision: int` -  specify what max precision a real number value can have

**Comment by rafaelrinaldi** [comment]
@jeddy3 The issue on `esformatter` project represents exactly what I meant. I still have mixed feelings about presets vs. plugins (another set of properties/values based on the context).

I think that the main project should only care about CSS itself and let alone naming approaches and stuff like that. In my opinion they should live on their own context.

Of course there are some drawbacks on this approach, for example, how are we going to make sure no one writes something that already exists (or it's similar) on another plugin? One way would be to discuss every new plugin and use the community to make that "parse", check if nothing being added is already working in another plugin.

I really don't know, there are trade offs...

**Comment by MoOx** [comment]
For my concern I will not spend time on preprocessor (but I will be open to contribution).
That being said I plan to use postcss so not sure if that will be easy.

Will take a deeper look into that project after releasing cssnext v1.0

**Comment by davidtheclark** [comment]
Following up on @jeddy3's comment ----- I'm actually thinking that the hard part of this project will be setting up the infrastructure, not writing the rules. I think there are already plenty of good rule ideas above, and what might be the important next step is to decide what the linter itself needs to do --- what API it will offer to the rule modules. Given the relative simplicity of CSS and the powers offered by PostCSS (parsing is already handled), I bet that once the core means of reporting errors is in place writing rules could be done fairly quickly.

I'm wondering *how simple* we can make the internals. Maybe *very* simple, right? It needs to be able to report a list of errors -- so needs to provide some way for rules to register errors. It might need to provide some shared helpers -- those would be identified as rules develop. I mean .... is that it? Or is there more I'm missing?

**Comment by MoOx** [comment]
Indeed it should not be a big deal to prepare core.
I've planned to start working on that in February, after cssnext 1.0 release.

**Comment by davidtheclark** [comment]
Ok. Sounds like you already have a plan in mind @MoOx. Let me know if there's anything I can help with.

**Comment by jeddy3** [comment]
@davidtheclark . Yeap, plenty of ideas for the rules. I thought I'd pull together what we have so far to make it easier to reference, now that @MoOx is soon to move onto the core and infrastructure.

Two things before I start though:

.1 

> @jeddy3 the problem for -before/after with custom string will be general indentation of the code

```css
.A { /* {: before " ", after "\n" */

}  { /* }: before "\n", after "\n\n" */

  .A-b {  { /* {: before " ", after "\n" */

  }  { /* }: before "\n" + INDENTATION, after "\n" */
```

> Here you might see a problem. Should we accept that some rules might rely on some others? (In this case ...brace-before, using indentation(-*).

I understand this example now! :) And I agree that a *custom string*, with spaces/tabs following a newline char, for `*-before/after` would interfere with the `indentation` and `separation` rules. Perhaps the `*-before/after` rules should only accept spaces, tabs, or a newline? But it does seem that the `*-before/after` rules might have to rely on the `indentation` one. Shall we wait until the rules are being implemented and come back to this one?

.2 

Am I right in thinking that setting `false` for a `*-no-*` rule is the same as not turning that rule on to begin with? e.g.

* `eol-no-whitespace: true` - there *must not* be any whiteline at the end of a line
* `eol-no-whitespace: false` - there *can* be whitespace at the end of a line (equivalent to not turning the rule on)

Where as rules without `*-no-*` are explicit e.g.

* `eof-newline: true` - the file *must* end with a newline
* `eof-newline: false` - the file *must not* end with a newline

I've tweaked some of the rules below based on this assumption...

I also looked a little deeper into the specs after reading the CSS parts of these [scss guidelines](http://sass-guidelin.es/#syntax--formatting), and I've shuffled around some of the rule groupings a bit because of it. Hopefully these new groupings still make sense!

---

# Rules

## File rules

* `eof-newline`: `bool` Newline at the end of the file/input
  * `true` - the file/input *must* end with a newline
  * `false` - the file/input *must not* end with a newline

## Line rules

* `eol-no-whitespace`: `bool` Disallow whitespace at the end of line
  * `true` - there *must not* be any whitespace at the end of a line
  * `false` - there *can* be whitespace at the end of a line (equivalent to not turning the rule on)

* `indentation`: `object` Specify whitespace for indentation
```js
{
  "indentation": {
    "default": 1,
    "char": "\t", 
    "block": true,
    "rule-set-by-selector-class-bem-element": true,
    "rule-set-by-selector-class-bem-modifier": false,
    "rule-set-by-selector-pseudo-classes": 2,
    "rule-set-by-selector-pseudo-element": true
  }
}
```

* `separation`: `object` Specify whitespace for separation

```js
{
  "statement-separation": {
    "default": 1
    "comment-blocks": true // separate comments-blocks by one line
    "statements": true, // separate statements by one line
    "at-rule-import-statements": false, // unless an @import rule (overriding, but OK as within a single rule?)
    "nested-statements": false // and unless the statement is nested
  }
}
```

## String rules ([ref](http://www.w3.org/TR/css-cascade-3/#at-import))

* `string-quotes`: `"\""`, `"'"` Specify quotes around strings
  * `"\""` - strings *must* be doubled quoted
  * `"'"` - strings *must* be single quoted

## Number rules ([ref](http://www.w3.org/TR/css3-values/#number))

* `number-leading-zero`: `bool` Leading zero before number
  * `true` - a number *must* have a leading zero
  * `false` - a number *must not* have a leading zero
* `number-trailing-zero`: `bool` Trailing zero after number
  * `true` - a number *must* have a trailing zero
  * `false` - a number *must not* have a trailing zero
* `number-unitized-zero`: `bool` Units for zero numbers
  * `true` - a number zero *must* have a unit
  * `false` - a number zero *must not* have a unit
* `number-max-precision`: `int` Maximum number of digits after the "." in a number

## Function rules ([ref](http://www.w3.org/TR/css3-values/#functional-notation))

* `function-opening-parenthesis-after`: `string` Specify whitespace after the opening parenthesis e.g. `""`, `" "`, `"\t"`
  * `""` - there *must not* be any whitespace after the opening parenthesis
  * `" "` - there *must* be a single space after the opening parenthesis
  * `"\t"` - there *must* be a single tab after the opening parenthesis
* `function-closing-parenthesis-before`: `string` Specify whitespace before the closing parenthesis
  * ...
* `function-coma-before`: `string` Specify whitespace before coma
* `function-coma-after`: `string` Specify whitespace after coma
* `function-operator-before`: `string` Specify whitespace before operator
* `function-operator-after`: `string` Specify whitespace after operator

## URL rules ([ref](http://www.w3.org/TR/css3-values/#url))

* `url-quotes`: `"\""`, `"'"`, `""` Specify quotes around URLs (URLs, unlike strings, can be unquoted)
  * `"\""` - URLs *must* be doubled quoted
  * `"'"` - URLs *must* be single quoted
  * `""` - URLs *must* be unquoted

## Color rules ([ref](http://dev.w3.org/csswg/css-color-4/))

Note: `color-no-*` replaces `value-color-keywords-prefered`

* `color-no-named`: `bool` Disallow named colors
  * `true`: named colors *must not* be used for colors
  * `false`: named colors *can* be used for colors
* `color-no-hex`: `bool` Disallow hex colors
* `color-no-rgb`: `bool` Disallow rgb colors
* `color-no-rgba`: `bool` Disallow rgba colors
* `color-no-hsl`: `bool` Disallow hsl colors
* `color-no-hsla`: `bool` Disallow hsla colors
* `color-no-hwb`: `bool` Disallow hwb colors
* `color-no-device-cmyk`: `bool` Disallow device-cmyk colors
* `color-no-gray`: `bool` Disallow gray colors
* `color-no-color`: `bool` Disallow color function colors
* `color-no-invalid-hex`: `bool` Disallow invalid hex (???)
* `color-hex-shorthand`: `bool`
  * `true` - hex colors *must* be shorthand
  * `false` - hex colors *must not* be shorthand

## :root rules

* `root-no-standard-properties`: `bool` Disallow use of standard properties in :root
  * `true`: there *must not* be standard properties in `:root`
  * `false`: there *can* be standard properties in `:root`

## Rule-set rules

* `rule-set-no-empty`: `bool` Disallow empty rule-sets
  * `true`: rule-sets *must not* be empty
  * `false`: rule-sets *can* be empty
* `rule-set-no-single-line`: `bool` Disallow single line rule-sets
  * `true`: rule-sets *must not* be on a single line
  * `false`: rule-sets *can* be on a single line
* `rule-set-body-selector-background-property`: `bool` Require the `background` property to be set in rule-sets that contain the `body` selector

## At-rule rules

* `at-rule-no-vendor-prefix`: `bool` Disallow vendor prefixes in @rules
  * `true`: there *must not* be vendor prefixes in @rules
  * `false`: there *can* be vendor prefixes in @rules

## Custom media rules ([ref](http://dev.w3.org/csswg/mediaqueries-4/#custom-mq))

* `custom-media-pattern`: `string` Preset or regex pattern to use to check names are matching a given pattern

## Media query list rules ([ref](http://www.w3.org/TR/mediaqueries-4/#media-query-list))

* `media-query-list-comma-before`: `string` Specify whitespace before coma
* `media-query-list-comma-after`: `string` Specify whitespace after coma

## Media query rules

* `media-query-opening-parenthesis-before`: `string` Specify whitespace
* `media-query-closing-parenthesis-after`: `string` Specify whitespace

## Media feature rules

* `media-feature-colon-before`: `string` Specify whitespace
* `media-feature-colon-after`: `string` Specify whitespace
* `media-feature-range-operator-before`: `string` Specify whitespace
* `media-feature-range-operator-after`: `string` Specify whitespace

## Selector rules

* `selector-attribute-quotes`: `"\""`, `"'"`, `""` Specify quotes for attribute selectors
* `selector-pattern`: `string` Preset or regex pattern to use to check names are matching a given pattern e.g. eg: `"b_e-m"` or `"B-e--m"`)) or regex (eg: `/\.([a-z]+-)?[A-Z][a-z]+(-[a-z][a-zA-Z]+)(--[a-z][a-zA-Z]+)/` to match .org-Block-eleMent--modifier
* `selector-max-length`: `int` Specify the maximum length a selector should have (int, descendant + child max length)
* `selector-max-specificity`: `int` Specify the maximum weight a selector should have according to css selectors specificity
* `selector-no-vendor-prefix`: `bool` Disallow vendor prefixes in selectors
* `selector-no-id`: `bool` Disallow use of id in selectors e.g: `#id`
  * `true`: there *must not* be ids in selectors
  * `false`: there *can* be ids in selectors
* `selector-no-type`: `bool` Disallow use of type in selectors e.g: `div`
* `selector-no-universal`: `bool` Disallow use of universal selector in selectors e.g: `*`
* `selector-no-attribute`: `bool` Disallow use of attributes in selectors e.g: `[attr]`
* `selector-no-qualified`: `bool` Disallow use of qualified selectors e.g: `div#id`
* `selector-no-pseudo-class`: `bool` Disallow use of pseudo-class in selectors
* `selector-no-pseudo-element`: `bool` Disallow use of pseudo-element in selectors
* `selector-no-combinator`: `bool` Disallow use of combinator in selectors
* `selector-combinator-before`: `string` Specify what whitespace should be used before combinator
* `selector-combinator-after`: `string` Specify what whitespace should be used after combinator
* `selector-no-delimiter`: `bool` Disallow use of delimiter in selectors
* `selector-delimiter-before`: `string` specify what whitespace should be used before delimiter
* `selector-delimiter-after`: `string` Specify what whitespace should be used after delimiter
* `selector-no-descendant`: `bool` Disallow use of descendant separator in selectors

## Custom selector rules ([ref](http://dev.w3.org/csswg/css-extensions/#custom-selectors))

* `custom-selector-pattern`: `string` Regex pattern to use to check names are matching a given pattern

## Block rules

* `block-opening-brace-before`: `string` Specify whitespace (???)
* `block-opening-brace-after`: `string` Specify whitespace (???)
* `block-closing-brace-before`: `string` Specify whitespace (???)
* `block-closing-brace-after`: `string` Specify whitespace (???)
* `block-single-line-opening-brace-before`: `string` Specify whitespace (???)
* `block-single-line-opening-brace-after`: `string` Specify whitespace (???)
* `block-single-line-closing-brace-before`: `string` Specify whitespace (???)
* `block-single-line-closing-brace-after`: `string` Specify whitespace (???)

Perhaps, like necolas suggested, the `*-single-line-*` rules can be handled with options within `block-opening-brace-after-*`? e.g.

```js
{
  "block-opening-brace-after": {
    "single": " ",
    "multiple": "\n"
  }
}
```

## Declaration block rules

* `declaration-block-properties-order`: `1 dimension array`, `2 dimensions array` (to specify bundle), `string` (preset) - `"alphabetical"` Order of properties within a declaration block
* `declaration-block-trailing-semicolon`: `bool` Trailing semicolon at the end of a declaration block
  * `true` - there *must* be a trailing semicolon
  * `false` - there *must not* be a trailing semicolon
* `declaration-block-no-duplicate-properties`: `bool`
  * `true` - there *must not* be any duplicate properties
  * `false` - there *can* be duplicate properties
* `declaration-block-single-line-max-declaration-length`: `int` The maximum length of a single line declaration block
* `declaration-block-single-line-max-declarations`: `int` The maximum number of declaration within a single line declaration block
* `declaration-block-semicolon-before`: `string` Specify whitespace (???)
* `declaration-block-semicolon-after`: `string` Specify whitespace (???)
* `declaration-block-single-line-semicolon-before`: `string` Specify whitespace (???)
* `declaration-block-single-line-semicolon-after`: `string` Specify whitespace (???)

Again, perhaps the `*-single-line-*` rules can be handled with complex options within `declaration-block-semicolon-*`?

## Declaration rules

* `declaration-no-important`: `bool` Disallow the use of !important
* `declaration-bang-before`: `string` Specify whitespace
* `declaration-bang-after`: `string` Specify whitespace
* `declaration-coma-before`: `string` Specify what whitespace should be used before the coma e.g. `background-image: url(flower.png), url(ball.png), url(grass.png);`
* `declaration-coma-after`: `string` Specify what whitespace should be used after the coma 

## Unit rules

* `unit-blacklist`: `array` Disallowed units e.g. `["px", "pt", "cm"]`
* `unit-whitelist`: `array` Only allowed units e.g. `["%", "rem", "em"]`

## Property rules

* `property-no-vendor-prefix`: `bool` Disallow vendor prefixes in properties
* `property-blacklist`: `array` Disallowed properties e.g. `["tranform", "background-size"]`
* `property-whitelist`: `array` Only allowed properties e.g. `["height", "width", "font-size"]`
* `property-unit-blacklist`: `object` Disallowed units for specific properties e.g. `{"width": ["%"], "height": ["%"]}`
* `property-unit-whitelist`: `object` Only allowed units for specific properties e.g. `{"font-size": ["em", "rem"]}`

## Custom property rules

* `custom-property-no-outside-root`: `bool` Disallow custom properties outside of `:root` 
  * `true` - there *must not* be custom properties outside of `:root`
  * `false` - there *can* be custom properties outside of `:root`
* `custom-property-pattern`: `string` Regex pattern to use to check names are matching a given pattern

## Comment rules

* `comment-case`: `string` - `"lower"` | `"sentence"` Case of comments

---

I've made this list available as a [gist](https://gist.github.com/jeddy3/6ca5fa8196fb8d1a90fb) so anyone can grab the raw markdown if they need it.

Right, I think that is everything that has been discussed (and a couple of others I just thought of).  It's definitely not all correct, especially the `indentation`, `separation` and `*-before-after` rules, but it's nice to have what we've got so far listed for reference. 

Anything in that lot that doesn't make immediate sense? I think the naming conventions that have emerged are pretty solid, and they seem to work well for new rules that have been thought up.

So, erm, just in case it isn't obvious... I'm not suggesting all these rules should exist (I'm not even sure which are feasible with the postcss parser). I just wanted to jot everything down for reference. That and I'm not very good at JavaScript (although I'm looking forward to learning more of it through by looking through this project's and postcss's source), and so I wanted to help out in some other way as I know I'll find this tool very useful! :)

**Comment by jeddy3** [comment]
@tunnckoCore Hello. To answer to [your question](https://github.com/stylelint/stylelint/pull/5#issuecomment-99553606) . I think we agreed (somewhere in this thread) two things: 

1. the linter should be unopinionated, i.e. the user can turn on only the rules she/he wants to use. 
2. each rule can be configured e.g.  using your example:

```css
a {
  width: calc( 2px * 2 );
}
```

You could enforce that by:

```js
function-opening-parenthesis-after: " ", // space char
function-closing-parenthesis-before: " "
```

Or, if you don't like that style and prefer no spaces e.g.

```css
a {
  width: calc(2px * 2);
}
```

You could enforce that using:

```js
function-opening-parenthesis-after: "", // empty string
function-closing-parenthesis-before: ""
```

Or, if you don't want to enforce any style when it comes to functions, then you simply don't enable the rule to begin with.

But, like I mentioned in the other thread, we had only just started naming the rules and defining their purpose. So, any feedback and suggestions would be welcome. For example, it might be clearer to use integers for number of spaces (as per MoOx original list) e.g. `function-opening-parenthesis-after: 0` or `function-opening-parenthesis-after: 1`. Or, maybe these two rules should be combined into a single rule e.g. `function-parentheses`.

I hope that helps :)

> it would be good to write examples for them.

Agreed :) The rule list got pretty long in the end and it become pretty difficult to discuss all of it. I think the idea now is to start implementing individual rules and, if needed, we can discuss the details of each rule as we come to them.

**Comment by tunnckoCore** [comment]
@jeddy3 yea, all that is clear to me. Im just talking about that Im totally against to have that two rules at all. ;d that's my feedback about, all other is great.

**Comment by jeddy3** [comment]
Updated rule list to reflect decisions in #38 and #40 .

**Comment by davidtheclark** [comment]
Fantastic!

**Comment by jeddy3** [comment]
Updated list to reflect:

a. merging of `declaration-block` and `rule-set` rules.
b. the `*-empty-line-before` rules.

**Comment by jeddy3** [comment]
@MoOx @davidtheclark I've been thinking about the remaining rules and how some could be approached differently to consolidate and de-clutter things (if only to save us from writing more documentation :) ):

What do you think about consolidating the following:

* `color-no-rgb`: `bool` Disallow rgb colors
* `color-no-rgba`: `bool` Disallow rgba colors
* `color-no-hsl`: `bool` Disallow hsl colors
* `color-no-hsla`: `bool` Disallow hsla colors
* `color-no-hwb`: `bool` Disallow hwb colors
* `color-no-device-cmyk`: `bool` Disallow device-cmyk colors
* `color-no-gray`: `bool` Disallow gray colors
* `color-no-color`: `bool` Disallow color function colors

Into:

* `color-function-blacklist`: `array` Disallowed color functions e.g. `["rgba", "hsl", "hwb"]`
* `color-function-whitelist`: `array` Only allowed color functions e.g. `["gray", "color", "device-cmyk"]`

Which would give us:

* `color-function-blacklist`: `array` Disallowed color functions e.g. `["rgba", "hsl", "hwb"]`
* `color-function-whitelist`: `array` Only allowed color functions e.g. `["gray", "color", device-cmyk"]`
* `color-no-named`: `bool` Disallow named colors
  * `true`: named colors *must not* be used for colors
  * `false`: named colors *can* be used for colors
* `color-no-hex`: `bool` Disallow hex colors
* `color-no-invalid-hex`: `bool` Disallow invalid hex colors
* `color-hex-shorthand`: `bool`
  * `true` - hex colors *must* be shorthand
  * `false` - hex colors *must not* be shorthand

I've been trying to think of as many use-cases as possible and this combination of rules seems to cover them all, while also making sense from an implementation perspective (e.g. separating named/ident colours out of the function white/blacklist rules).

Secondly, I'm wondering about the value of the `selector-no-*`rules:

* `selector-no-id`: `bool` Disallow use of id in selectors e.g: `#id`
  * `true`: there *must not* be ids in selectors
  * `false`: there *can* be ids in selectors
* `selector-no-type`: `bool` Disallow use of type in selectors e.g: `div`
* `selector-no-universal`: `bool` Disallow use of universal selector in selectors e.g: `*`
* `selector-no-attribute`: `bool` Disallow use of attributes in selectors e.g: `[attr]`
* `selector-no-qualified`: `bool` Disallow use of qualified selectors e.g: `div#id`
* `selector-no-pseudo-class`: `bool` Disallow use of pseudo-class in selectors
* `selector-no-pseudo-element`: `bool` Disallow use of pseudo-element in selectors
* `selector-no-combinator`: `bool` Disallow use of combinator in selectors
* `selector-no-delimiter`: `bool` Disallow use of delimiter in selectors
* `selector-no-descendant`: `bool` Disallow use of descendant separator in selectors

I'm wondering if they'll actually be used and aren't they are already covered by:

* `selector-pattern`: `string` regex

I'm also wondering whether `selector-pattern` should just focus on these types of patterns and leave the more advanced bem pattern matching to postcss-bem-linter? Should be strip out the `selector-no-*`rules and simify `selector-pattern` to regex only rather than presets too?

What do you guys think about these two proposals?

**Comment by davidtheclark** [comment]
I'm in favor of the color blacklist/whitelist simplification.

Once @ben-eb has finished https://github.com/postcss/postcss-selector-parser, writing these `selector-no-*` patterns should be a breeze. For many people writing a custom regex that is actually accurate everywhere will be a serious problem; and given the attention to detail and testing that is going into `postcss-selector-parser`, I'd imagine that using it to evaluate selectors will almost always be more accurate than a user's regex.

So I'm in favor of keeping the `selector-no-*` rules in our list *and* the `selector-pattern`. That said, I think we should be able to build `selector-pattern` at any point, while for the `selector-no-*` rules we should just wait for the `postcss-selector-parser`.

Re: `selector-pattern` -- Is this pattern intended to match against selector list, complex, compound, of simple selectors? (http://dev.w3.org/csswg/selectors/#structure). Compound might be the most useful? That's what bem-linter is checking, I believe.

I'm not sure about `selector-no-delimiter`. Do we think anybody would turn that on?

**Comment by jeddy3** [comment]
> I'm in favor of the color blacklist/whitelist simplification.

Sweet. I've updated the list accordingly.

> For many people writing a custom regex that is actually accurate everywhere will be a serious problem

Good point.

> So I'm in favor of keeping the selector-no-* rules in our list and the selector-pattern. That said, I think we should be able to build selector-pattern at any point, while for the selector-no-* rules we should just wait for the postcss-selector-parser.

Agreed.

> Compound might be the most useful? 

Agreed. Should we call the rule `selector-compound-pattern` to be clear?

> I'm not sure about selector-no-delimiter. Do we think anybody would turn that on?

I can't think of a use-case. I'll remove it for now and if a request pops up later down the line we can add it back in.

**Comment by MoOx** [comment]
Like @davidtheclark I think you color proposal is nice.
 I also think that we should stick with selector-no-* like he explained.

I will probably will be the first to use `selector-no-delimiter` when using strict bem: when you use classes only, you never use delimiters am I right ?

`selector-compound-pattern` make sense but `selector-pattern` could cover entire selector. Different use cases I think.

**Comment by jeddy3** [comment]
> I will probably will be the first to use selector-no-delimiter when using strict bem: when you use classes only, you never use delimiters am I right ?

I'm not totally sure how strict bem works, but I've added the rule back in anyway :)

> selector-compound-pattern make sense but selector-pattern could cover entire selector. Different use cases I think.

Good point. Perhaps we should be really specific here and replace `selector-pattern` with:

* `selector-simple-pattern`
* `selector-compound-pattern`
* `selector-complex-pattern`
* `selector-list-pattern`

Does each have a different use case? Also, are we stepping on the toes of all the other `selector-*` rules by doing this e.g. `selector-compound-pattern` steps on the toes of:

* `selector-no-combinator`
* `selector-combinator-space-before|after`

I don't think we need to decide on this now. I think we can wait to see what approach will be easiest when using `postcss-selector-parser`. Sound good?

**Comment by MoOx** [comment]
Sounds good

**Comment by ben-eb** [comment]
@davidtheclark I think you would be best to use `eachInside` for these. I wrote some methods on top of it to filter the nodes based on type, but implementing something like `selector-no-universal` is this trivial:

```js
parser.eachInside(function (selector) {
  if (selector.type === 'universal' && opts.noUniversal) {
    // ... your logic for warning the user
  }
});
```

I'm going to rework this iterator to be safe for removing nodes whilst iterating, like PostCSS does, but the parser itself is pretty solid right now. Just need more tools to work with the AST I reckon. :smile:

**Comment by davidtheclark** [comment]
@ben-eb Awesome. Just let us know when you think it's finalized enough for us to work with.

@MoOx I guess I don't understand why you would not use delimiters because you're using classes. Couldn't you have something like:

```css
.my-component__thing--odd,
.my-component__otherpart {
  color: pink;
}

.my-component__thing--even {
  background: oragen;
}
```

I don't recall seeing a rule that you must have only one selector per rule -- maybe I missed it?

**Comment by ben-eb** [comment]
@davidtheclark I just published `0.0.5` on npm which has safe AST iteration on `each` and `eachInside` which I think should be enough to work with for your use case. No API docs yet but I'll be working on those and looking to publish a `1.0.0` soon with usage instructions - for now just check the tests directory which has tons of examples. Also the sooner you start working with the module, you can identify issues before I stabilise the API. :smile:

**Comment by ben-eb** [comment]
@davidtheclark Current API: https://github.com/postcss/postcss-selector-parser/blob/master/API.md - if you have any suggestions, please let me know. https://github.com/postcss/postcss-selector-parser/issues/5

**Comment by davidtheclark** [comment]
Thanks @ben-eb. At some point here I'll have a few minutes and I'll try to write some selector linting rules to give it a shot.

**Comment by jeddy3** [comment]
Updated list to reflect decisions in #178

**Comment by davidtheclark** [comment]
@jeddy3 : Because of insupportable length of this issue, I suggest that we close it and open new issues for any of the rules listed way up above that we definitely want to implement or urge fantasy contributors to implement. What do you think?

**Comment by jeddy3** [comment]
@davidtheclark Makes sense to me. I admit I'll be a little sad to see it go though as it's been with us since the start. Kind of like saying good bye to a comfy sofa that you've had for years, but has out-lived its usefulness :) 

I'll pick out the ones I'd like to see. Then, if you can do the same then close the issue.

**Comment by jeddy3** [comment]
I'm done :)

**Comment by davidtheclark** [comment]
Thanks.
